planner/implementation/simple_plans.go

```diff
@@ -144,3 +144,28 @@ func (impl *TiDBTopNImpl) CalcCost(outCount float64, children ...memo.Implementa
 func NewTiDBTopNImpl(topN *plannercore.PhysicalTopN) *TiDBTopNImpl {
 	return &TiDBTopNImpl{baseImpl{plan: topN}}
 }
+
+// UnionAllImpl is the implementation of PhysicalUnionAll.
+type UnionAllImpl struct {
+	baseImpl
+}
+
+// CalcCost implements Implementation CalcCost interface.
+func (impl *UnionAllImpl) CalcCost(outCount float64, children ...memo.Implementation) float64 {
+	var childMaxCost float64
+	for _, child := range children {
+		childCost := child.GetCost()
+		if childCost > childMaxCost {
+			childMaxCost = childCost
+		}
+	}
+	selfCost := float64(1+len(children)) * impl.plan.SCtx().GetSessionVars().ConcurrencyFactor
+	// Children of UnionAll are executed in parallel.
+	impl.cost = selfCost + childMaxCost
```

Comment:
UnionAll does not satisfy the constriant `cost = selfCost + childCost`, either. I think `ScaleCostLimit` is not powful enough. I will remove `ScaleCostLimit` and introduce a new method for `cost limit` in next PR.